### ISSUE TITLE
Feature Request: Add Gaussian Noise 

### ISSUE THREAD

**Issue by camilleconcepcion3012** [RobotecAI/RGLGazeboPlugin#29]
Hi sir, is it possible to add a gaussian noise feature to this plugin? I am using this plugin for a project and wanted to add gaussian noise to the point cloud data. I did some reseacrh and found that there is an API call within the RobotecGPULidar github that generates a node to add gaussian noise to the point cloud data. I have tried to integrate the feature to the plugin by upgrading the core.h file to v0.13.1. However, I ran into the issue where the library libRobotecGPULidar.so provided clashes with the upgraded core.h file, presumably due to in being from v0.11.3. I was wondering if there is an updated version of the libRobotecGPULidar.so for v0.13.1. Thanks in advance.

**Comment by PawelLiberadzki** [comment]
Hi @camilleconcepcion3012, I am not sure what was exactly the process in your case - can you maybe elaborate? Binaries in releases are the newest that we support. We did not prepare or publish more up-to-date version.

However, if you are really willing to put some work into it, your goal should be possible to achieve. You have to build this plugin from source following [this](https://github.com/RobotecAI/RGLGazeboPlugin?tab=readme-ov-file#building-from-source) paragraph. Before building you have to update RGL_VERSION (for binary) and RGL_COMMIT (for core.h) [here](https://github.com/RobotecAI/RGLGazeboPlugin/blob/main/RGLServerPlugin/CMakeLists.txt). For v0.13.1 it should be:
`set(RGL_VERSION "0.13.1")`
`set(RGL_COMMIT "9786a9689ee0c84208753b98ee35f2c7ddb67d24")`

This should give you access to v0.13.1 RGL API calls. Then, you should create gaussian noise node [here](https://github.com/RobotecAI/RGLGazeboPlugin/blob/main/RGLServerPlugin/src/Lidar.cc#L87) and connect it to RGL graph [here](https://github.com/RobotecAI/RGLGazeboPlugin/blob/main/RGLServerPlugin/src/Lidar.cc#L97). You can find more information about this node [here](https://github.com/RobotecAI/RobotecGPULidar/blob/develop/docs/GaussianNoise.md). RGL API rather did not change much or at all between v0.11.3 and v0.13.1 - if you will have any problems, you will have to update RGL native calls probably mostly in [this](https://github.com/RobotecAI/RGLGazeboPlugin/blob/main/RGLServerPlugin/src/Lidar.cc) file.

Best regards!

**Comment by camilleconcepcion3012** [comment]
Hi @PawelLiberadzki, thanks for the response!

My process is quite similar to the one you have suggested. I defined the gaussian noise node in the RGLServerInstance.hh via this code:
![image](https://github.com/RobotecAI/RGLGazeboPlugin/assets/106718143/c19b1f9f-2871-491c-a352-1cfbf9ab52ac)

After which, I created the gaussian noise node and connected it to the RGL graph as shown in the code below(for now using hard-coded values):
![image](https://github.com/RobotecAI/RGLGazeboPlugin/assets/106718143/93ea7770-e1aa-4b33-9d8b-fccd822be78a)

Then, I updated the RGL_VERSION and RGL_COMMIT to 'v0.13.1' and '9786a9689ee0c84208753b98ee35f2c7ddb67d24' and proceeded to build the plugin and copied the updated binaries to my ign_gazebo_6 plugin directory. However, when running the sonoma test world, it resulted in the following error:
![image](https://github.com/RobotecAI/RGLGazeboPlugin/assets/106718143/565e6781-e54b-446a-b05b-8b1abdb2c493)

Further troubleshooting, I discovered that the dowloaded binaries are empty:
![image](https://github.com/RobotecAI/RGLGazeboPlugin/assets/106718143/8952188e-c659-4ecb-9119-e358e50d99d5)

Could you please advise as to how I can solve this issue?

**Comment by msz-rai** [comment]
Hi @camilleconcepcion3012
We have missed that since v0.12.0, RGL binaries are packed into zip files. Sorry for that.

You have two options:
1. Download the zip file ([link](https://github.com/RobotecAI/RobotecGPULidar/releases/tag/v0.13.1)) to the `RGLServerPlugin/include/rgl` and extract it manually.
2. Modify [CMakeLists](https://github.com/RobotecAI/RGLGazeboPlugin/blob/main/RGLServerPlugin/CMakeLists.txt) to make it done automatically. We have already implemented a similar solution in the other of our plugins ([link](https://github.com/RobotecAI/o3de-rgl-gem/blob/development/Code/FindRGL.cmake)). I think this may be helpful.

It's also worth knowing that RGL binaries for our releases are built with all available extensions. For instance, with v0.12.0 there was a new extension added (ROS2), so to run RGL binary from the release v0.12.0+ you have to provide ROS2 libraries. You can do this in two ways:
1. Install ROS2 on your machine and source it before running the simulation.
2. Copy all libraries from the zip file downloaded from the release into the same location as `libRobotecGPULidar.so`. The zip file contains all of the dependent libraries (mostly ROS2 libraries) required for RGL.

Alternatively, you can build RGL yourself from source without ROS2 extension. Instructions can be found in the [RGL README](https://github.com/RobotecAI/RobotecGPULidar/blob/develop/README.md).

Best regards!

**Comment by camilleconcepcion3012** [comment]
Hi @msz-rai
I followed your 2nd method and copied all the libraries and RGL binaries to the same location. However, when building the plugin, I encountered a new error:

![Screenshot from 2024-02-28 08-59-47](https://github.com/RobotecAI/RGLGazeboPlugin/assets/106718143/864f7d6b-3d5b-4381-a9df-b1153c5e9434)


I believe this is due to my environment being on Ubuntu 20.04. For project-specific reasons, I am unable to update to Ubuntu 22.04 and thus trying the alternative method which is to build RGL from source on my current environment. I have followed the steps in the RGL README but I am still having trouble building RGL due to configurations in the CMakeLists.txt. Could you please advise as to how I can solve this issue?

**Environment**
- Ubuntu: 20.04
- Nvidia Optix: 7.2
- Nvidia Driver: 520
- CUDA: V11.8.89

**Comment by msz-rai** [comment]
@camilleconcepcion3012 
> I followed your 2nd method and copied all the libraries and RGL binaries to the same location. However, when building the plugin, I encountered a new error (...). I believe this is due to my environment being on Ubuntu 20.04

You are correct. The released RGL binaries are built with ROS2 Humble so it is supposed to be used on Ubuntu22. 

> I have followed the steps in the RGL README but I am facing this error when building:

We are not supporting RGL on Ubuntu20 actively, so we cannot ensure proper work on it. However, the error message you got is related to one of our internal tools. You can try to build RGL without those tools with the following command:
```bash
./setup.py --cmake="-DRGL_BUILD_TOOLS=OFF"
```

I hope it will solve your issue.

Best regards!

**Comment by msz-rai** [comment]
@camilleconcepcion3012 

Good news: the Plugin has been updated to the newer RGL version (0.17.0). This makes the Gaussian noise feature should be added without requiring any custom RGL builds on your side. If you haven't resolved your issue yet, the latest changes to the Plugin might help you move forward.
